### ISSUE TITLE
[ FEATURE ] Parse invalid token

### ISSUE THREAD

**Issue by 6562680** [firebase/php-jwt#347]
Currently library throws exception on decode if token is invalid.
Sometimes you need to get payload of the token to refresh it with old data
So token is valid but expired or not-before invalid

How about non-runtime exceptions or maybe additional method "parse" that you call if exception was thrown

Thank you

**Comment by cottton** [comment]
BIG NO from me.

TL;DR: let the auth server create a new token.

See https://github.com/firebase/php-jwt/issues/119

You should never ever re-use the payload of an invalid token.

Q: "The token is just expired. I know from the code that the signature has been checked before, so i can use the payload."
A: And what if the code gets a change so the signature validation gets moved below the expired-check?
You COULD check on EVERY new release if the code is still the same, but believe me: you will not =)

Q: "Yea, what ever, ... what if i really know that the token payload has not been modified?"
A: Then you may use outdated payload which f.e. could allow a user access to something he actually should not get access to anymore.
Or simple things like changed email, name, ... wont make it into the token.

---

If you got an invalid token, 
and you do not have access to data needed to create the payload
then you are not in  the right place to create the token.

Example: 
is the token on a mobile app expired?
Forward to login.

**Comment by 6562680** [comment]
I agree with your notes, but, please, listen to me
The token stores user data that could (but not should) be in the database. Of course usually token its just a mirror of maybe roles/permissions/user_name/full_name/avatar of the user.

While refreshing we can get actual values to create new token.

How about JWT could be used to deliver session between backends ?

**Comment by cottton** [comment]
1. thats to general for me. What you mean by "deliver sessions" ?
Do you mean the JWT contains the session id ? Or the session content?
In case session id: why? You could use the id in the header. Why extra packing it into a header with a auth token (JWT).
In case session content: _i dont know any example where this would make sense. So - why :)_

2. You did not read the https://github.com/firebase/php-jwt/issues/119
(direct link to example post: https://github.com/firebase/php-jwt/issues/119#issuecomment-525464922)

Auth server creates both token - access token and refresh token.
Auth server does not store the access token.
Auth server does store the refresh token.
_[...]_

And the refresh token (IMO) SHOULD be stored like

    {refresh_token} => {user_id} // Able to find user by refresh token
    {user_id} => {refresh_token (array)} // Able to find refresh token(s) by user *1.
 
    *1) f.e. to delete the refresh token on "logout" request. Client calls "logout" with current access token. 
    Access token is valid, user id is in payload, and server now can "logout" the user .
    Or admin just needs to "logout" a user.
    (If token is invalid then no "logout" ofc. Forward to login. Nobody should be able to "logout" random users.)
    (Note: "logout" in quotes, because of JWT design. The access token is still valid out there until it expires.)

So you can call the auth server with the refresh token alone to receive a new access token (and refresh token ofc - because one time refresh token).

---

**In details:** 

_TL; DR: if you handle the token for the client then store its user id from the beginning, so you do not rely on the invalid token._ 

What i **think** you do by

> i always write damn split, atob, JSON.parse() to get user id

you have a client app that represents the frontend.
User -> browser -> frontend app -> controller
and here, in the controller you seem to use a cookie session to get frontend app and browser "in touch".

In this or similar case you build a "bridge" for the user, which is ok (ish).
But you got everything from the beginning:
- user login (user, pw)
- direct call to frontend app
- fronend app internal call to auth server
- auth server (on success) returns both token
- frontend app receives both token
- frontend app stores both token **for** the user (i.e. in a cookie session)
(where the session data ofc stays on fronend app (server-) side, and the user receives a session cookie id)
- frontend app returns the success response with the session cookie
- ... 
- user requests something "internal" (that the frontend app does not own)
- frontend app receives request with session cookie
- frontend app checks the session cookie (ofc) and finds the token
- frontend app "forwards" the request with the access token 
(so frontend requests an internal app **as** user, with his token)

Now the access token has expired.
- fronend app requests a new access token  **for** the user, and stores the new access and refresh token
(User does not need to know about that. He is valid as long his session cookie is valid)

Now with both token expired.
- fronend app forwards to (fronend-) login (where the hole process starts over).

Now the session cookie has expired.
- same as if both token are expired - send him to (frontend-) login.

---

This could work, but also may makes it complicated.
You could just let the user store its (both) token him self (JS framework -> save local storage -> ...). 
It depends ... it could be easier with a "bridge". F.e. if all the token logic is already written on server side,
or if you do not use a JS framework at all.

**Comment by wuuti** [comment]
How about the following use case: your application receives tokens, which are considered to be invalid (e.g. a wrong signature). As this _could_ be an intrusion attempt, you may want to log the decoded payload/body of the JWT, or use data from the payload in a monitoring tool / metric to see e.g. certain platforms/users/apps (whatever your resource owner or claims are in your business case) always sending expired tokens.

**Comment by zagreusinoz** [comment]
Or (my use case) you want to retrieve a JTI to add that to a blacklist.

**Comment by bshaffer** [comment]
It's relatively trivial to decode a token without validation, if you so require. We will not encourage such behavior in this library.
